pkgs/build-support/mkshell/default.nix

```
@@ -28,18 +28,63 @@ let
     "propagatedNativeBuildInputs"
     "shellHook"
   ];
+
+  merged = {
+    buildInputs = mergeInputs "buildInputs";
+    nativeBuildInputs = packages ++ (mergeInputs "nativeBuildInputs");
+    propagatedBuildInputs = mergeInputs "propagatedBuildInputs";
+    propagatedNativeBuildInputs = mergeInputs "propagatedNativeBuildInputs";
+  };
+
+  docPathShellHook = let
+    unwrapCC = pkg: if pkg ? cc then unwrapCC pkg.cc else pkg;
+    docInputs = builtins.concatLists (builtins.attrValues merged)
+                # grab the unwrapped cc, as wrappers may or may not
+                # retain doc outputs
+                ++ (lib.optional (stdenv ? cc) (unwrapCC stdenv.cc))
+                ++ stdenv.defaultBuildInputs ++ stdenv.defaultNativeBuildInputs
+                ++ stdenv.initialPath;
+    docDrvs = lib.unique (builtins.filter lib.isDerivation docInputs);
+    docPaths = type: lib.pipe docDrvs [
+      (map (lib.getOutput type))
+      (map (out: builtins.toPath "${out}/share/${type}"))
+      (builtins.filter builtins.pathExists)
+    ];
+    infoPaths = docPaths "info";
+    manPaths = docPaths "man";
+
+    # some packages neglect to build an info dir file
+    infoDirAddenda = stdenv.mkDerivation {
+      inherit (stdenv) system;
+      name = "${name}-info-dir-addenda";
+      passAsFile = ["buildCommand"];
+      buildCommand = ''
+        shopt -s nullglob
+        mkdir -p $out/share/info
+        for path in ${builtins.concatStringsSep " " infoPaths}; do
+          [[ -s $path/dir ]] && continue || :
+          for file in $path/*.info{,.gz}; do
+            ${buildPackages.texinfo}/bin/install-info $file $out/share/info/dir
+          done
+        done
+      '';
+    };
+  in ''
+    export INFOPATH=${builtins.concatStringsSep ":" (infoPaths ++ [(builtins.toPath "${infoDirAddenda}/share/info")])}:$INFOPATH
+    export MANPATH=${builtins.concatStringsSep ":" manPaths}:$MANPATH
```

If you do this for `MANPATH` and `INFOPATH` then you don't need to add the `/share/${type}` suffix in `docPaths`
```suggestion
    export MANPATH=${lib.makeSearchPath "share/man" manPaths}:$MANPATH
```